Query: Include collection and order by subquery throws 'Variable names must be unique within a query batch or stored procedure.'

Tasteful opened the pull request:
Potential fix for #4521, the fix is working in my local application but cant run all the tests because of #5395 that will fail all orderby tests.

anpete commented on src/Microsoft.EntityFrameworkCore.Specification.Tests/IncludeTestBase.cs:
```diff
@@ -490,6 +490,26 @@ var customers
         }
 
         [Fact]
+        public virtual void Include_collection_order_by_subquery()
```
@Tasteful Can you add the derived Sql Server version of this so we can see the SQL?

Tasteful commented:
Is now generating the follwoing SQL
```
exec sp_executesql N'SELECT TOP(1) [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE [c].[CustomerID] = @__customerId_0
ORDER BY (
    SELECT TOP(1) [oo].[OrderDate]
    FROM [Orders] AS [oo]
    WHERE [c].[CustomerID] = [oo].[CustomerID]
), [c].[CustomerID]',N'@__customerId_0 nvarchar(4000)',@__customerId_0=N'ALFKI'
go
exec sp_executesql N'SELECT [o].[OrderID], [o].[CustomerID], [o].[EmployeeID], [o].[OrderDate]
FROM [Orders] AS [o]
INNER JOIN (
    SELECT DISTINCT TOP(1) (
        SELECT TOP(1) [oo].[OrderDate]
        FROM [Orders] AS [oo]
        WHERE [c].[CustomerID] = [oo].[CustomerID]
    ) AS [c0_0], [c].[CustomerID]
    FROM [Customers] AS [c]
    WHERE [c].[CustomerID] = @__customerId_0
    ORDER BY [c0_0], [c].[CustomerID]
) AS [c0] ON [o].[CustomerID] = [c0].[CustomerID]
ORDER BY [c0_0], [c0].[CustomerID]',N'@__customerId_0 nvarchar(4000)',@__customerId_0=N'ALFKI'
go
```

Original failing SQL looks like
``` 
exec sp_executesql N'SELECT TOP(1) [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE [c].[CustomerID] = @__customerId_0
ORDER BY (
    SELECT TOP(1) [oo].[OrderDate]
    FROM [Orders] AS [oo]
    WHERE [c].[CustomerID] = [oo].[CustomerID]
), [c].[CustomerID]',N'@__customerId_0 nvarchar(4000)',@__customerId_0=N'ALFKI'
go
exec sp_executesql N'SELECT [o].[OrderID], [o].[CustomerID], [o].[EmployeeID], [o].[OrderDate]
FROM [Orders] AS [o]
INNER JOIN (
    SELECT DISTINCT TOP(1) (
        SELECT TOP(1) [oo].[OrderDate]
        FROM [Orders] AS [oo]
        WHERE [c].[CustomerID] = [oo].[CustomerID]
    ), [c].[CustomerID]
    FROM [Customers] AS [c]
    WHERE [c].[CustomerID] = @__customerId_0
    ORDER BY (
        SELECT TOP(1) [oo].[OrderDate]
        FROM [Orders] AS [oo]
        WHERE [c].[CustomerID] = [oo].[CustomerID]
    ), [c].[CustomerID]
) AS [c0] ON [o].[CustomerID] = [c0].[CustomerID]
ORDER BY (
    SELECT TOP(1) [oo].[OrderDate]
    FROM [Orders] AS [oo]
    WHERE [c].[CustomerID] = [oo].[CustomerID]
), [c0].[CustomerID]',N'@__customerId_0 nvarchar(4000)',@__customerId_0=N'ALFKI'
go
```

Tasteful commented:
After this change also the attached test in #5038 works.

Tasteful commented:
After this change the test from #5038 is working for SqlLite and SqlServer but not InMemory.

smitpatel commented:
Yes this is a fix for #5038 too. One of the solutions of #5038 is to assign an alias to the subquery in projection and order by the alias. That way SqlServer knows that all order by columns appear in projection.

@Tasteful - What is the error you are getting for InMemory? Any stack trace?

Tasteful commented:
@smitpatel The error for the inmemory test are the following
```
Test Name:	Microsoft.EntityFrameworkCore.InMemory.FunctionalTests.IncludeInMemoryTest.Include_collection_order_by_collection_column
Test FullName:	Microsoft.EntityFrameworkCore.InMemory.FunctionalTests.IncludeInMemoryTest.Include_collection_order_by_collection_column
Test Source:	 : line -1
Test Outcome:	Failed
Test Duration:	0:00:00,436

Result StackTrace:	
at Xunit.Assert.NotNull(Object object) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.assert\Asserts\NullAsserts.cs:line 15
   at Microsoft.EntityFrameworkCore.Specification.Tests.IncludeTestBase`1.Include_collection_order_by_collection_column() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Specification.Tests\IncludeTestBase.cs:line 468
Result Message:	Assert.NotNull() Failure
```
But the test from #5038 is wrong, the `.Where(c => c.CustomerID.StartsWith("w"))` does not exists, changed to `.Where(c => c.CustomerID.StartsWith("W"))` and then I got
```
Test Name:	Microsoft.EntityFrameworkCore.InMemory.FunctionalTests.IncludeInMemoryTest.Include_collection_order_by_collection_column
Test FullName:	Microsoft.EntityFrameworkCore.InMemory.FunctionalTests.IncludeInMemoryTest.Include_collection_order_by_collection_column
Test Source:	 : line -1
Test Outcome:	Failed
Test Duration:	0:00:00,361

Result StackTrace:	
at System.Collections.Comparer.Compare(Object a, Object b)
   at System.Linq.CachingComparer`2.Compare(TElement element, Boolean cacheLower)
   at System.Linq.OrderedEnumerable`1.TryGetFirst(Boolean& found)
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
   at lambda_method(Closure , QueryContext )
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass19_1`1.<CompileQuery>b__1(QueryContext qc) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\QueryCompiler.cs:line 168
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\QueryCompiler.cs:line 82
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\EntityQueryProvider.cs:line 37
   at Microsoft.EntityFrameworkCore.Specification.Tests.IncludeTestBase`1.Include_collection_order_by_collection_column() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Specification.Tests\IncludeTestBase.cs:line 461
Result Message:	At least one object must implement IComparable.
```
because of `.OrderByDescending(c => c.Orders.OrderByDescending(oo => oo.OrderDate).FirstOrDefault())` that will return the `Order` that not is IComparable.

Changed the orderby to `.OrderByDescending(c => c.Orders.OrderByDescending(oo => oo.OrderDate).FirstOrDefault().OrderDate)` and then the test pass.

Tasteful commented:
@smitpatel @anpete I still have problem to run all the OrderBy test (blocked by #5395) so have total 66 failing and 4 of them (regarding to error) is not related to the above bug.

It's the sync/async version of the following errors.
```
Test Name:	Microsoft.EntityFrameworkCore.Sqlite.FunctionalTests.IncludeAsyncSqliteTest.Include_duplicate_collection_result_operator
Test FullName:	Microsoft.EntityFrameworkCore.Sqlite.FunctionalTests.IncludeAsyncSqliteTest.Include_duplicate_collection_result_operator
Test Source:	 : line -1
Test Outcome:	Failed
Test Duration:	0:00:01,835

Result StackTrace:	
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.<MoveNext>d__5.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\AsyncLinqOperatorProvider.cs:line 146
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.<ToListAsync>d__129`1.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\EntityFrameworkQueryableExtensions.cs:line 2137
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Specification.Tests.IncludeAsyncTestBase`1.<Include_duplicate_collection_result_operator>d__31.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Specification.Tests\IncludeAsyncTestBase.cs:line 525
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass46_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 227
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Linq.AsyncEnumerableExtensions.AsyncSelectEnumerable`2.AsyncSelectEnumerator.<MoveNext>d__3.MoveNext() in C:\Git\EntityFramework\src\Shared\AsyncEnumerableExtensions.cs:line 47
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.<IncludeAsync>d__13.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\QueryBuffer.cs:line 219
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.EntityFrameworkCore.Query.AsyncQueryMethodProvider.<>c__DisplayClass33_0`1.<<_Include>b__1>d.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\AsyncQueryMethodProvider.cs:line 503
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Linq.AsyncEnumerableExtensions.AsyncSelectEnumerable`2.AsyncSelectEnumerator.<MoveNext>d__3.MoveNext() in C:\Git\EntityFramework\src\Shared\AsyncEnumerableExtensions.cs:line 52

   at Microsoft.Data.Sqlite.Interop.MarshalEx.ThrowExceptionForRC(Int32 rc, Sqlite3Handle db)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
   at Microsoft.Data.Sqlite.SqliteCommand.<ExecuteDbDataReaderAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.<ExecuteAsync>d__20.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Storage\Internal\RelationalCommand.cs:line 344
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable.AsyncEnumerator.<MoveNext>d__8.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\Internal\AsyncQueryingEnumerable.cs:line 74
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncIncludeCollectionIterator.RelatedValuesEnumerable.RelatedValuesEnumerator.<MoveNext>d__2.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\Internal\AsyncIncludeCollectionIterator.cs:line 61
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at System.Linq.AsyncEnumerableExtensions.AsyncSelectEnumerable`2.AsyncSelectEnumerator.<MoveNext>d__3.MoveNext() in C:\Git\EntityFramework\src\Shared\AsyncEnumerableExtensions.cs:line 47
Result Message:	
One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (SQLite Error 1: 'no such column: t0.CustomerID0'.))))
One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (SQLite Error 1: 'no such column: t0.CustomerID0'.)))
One or more errors occurred. (One or more errors occurred. (SQLite Error 1: 'no such column: t0.CustomerID0'.))
One or more errors occurred. (SQLite Error 1: 'no such column: t0.CustomerID0'.)
SQLite Error 1: 'no such column: t0.CustomerID0'.
```

And 
```
Test Name:	Microsoft.EntityFrameworkCore.Sqlite.FunctionalTests.IncludeSqliteTest.Include_duplicate_collection_result_operator
Test FullName:	Microsoft.EntityFrameworkCore.Sqlite.FunctionalTests.IncludeSqliteTest.Include_duplicate_collection_result_operator
Test Source:	 : line -1
Test Outcome:	Failed
Test Duration:	0:00:01,827

Result StackTrace:	
at Microsoft.Data.Sqlite.Interop.MarshalEx.ThrowExceptionForRC(Int32 rc, Sqlite3Handle db)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMethod, IReadOnlyDictionary`2 parameterValues, Boolean openConnection, Boolean closeConnection) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Storage\Internal\RelationalCommand.cs:line 222
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues, Boolean manageConnection) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Storage\Internal\RelationalCommand.cs:line 95
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable.Enumerator.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\Internal\QueryingEnumerable.cs:line 69
   at Microsoft.EntityFrameworkCore.Query.Internal.IncludeCollectionIterator.<GetRelatedValues>d__4.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\Internal\IncludeCollectionIterator.cs:line 28
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.Include(QueryContext queryContext, Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedEntitiesLoaders, Int32 currentNavigationIndex, Boolean queryStateManager) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\QueryBuffer.cs:line 160
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.Include(QueryContext queryContext, Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedEntitiesLoaders, Boolean queryStateManager) in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\QueryBuffer.cs:line 134
   at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_Include>d__30`1.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\QueryMethodProvider.cs:line 329
   at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_Include>d__30`1.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Relational\Query\QueryMethodProvider.cs:line 327
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.<_TrackEntities>d__15`2.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\LinqOperatorProvider.cs:line 149
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore\Query\Internal\LinqOperatorProvider.cs:line 125
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Specification.Tests.IncludeTestBase`1.Include_duplicate_collection_result_operator() in C:\Git\EntityFramework\src\Microsoft.EntityFrameworkCore.Specification.Tests\IncludeTestBase.cs:line 676
Result Message:	SQLite Error 1: 'no such column: t0.CustomerID0'.
```

smitpatel commented:
@Tasteful - I think I know what had caused earlier error for InMemory. string search is case sensitive for `InMemory` but not for `SQLite` or `SqlServer`. So they return different number of results.
For the last error, can you check the generated SQL. Looks like some invalid column is being used in SQL.

Tasteful commented:
@smitpatel 
It's the `ORDER BY [t].[CustomerID], [t0].[CustomerID0]` that either should be `ORDER BY [t].[CustomerID], [CustomerID0]` or `ORDER BY [t].[CustomerID], [t0].[CustomerID]`.

Is it best to use the alias or the `t0`-column?

smitpatel commented:
Looking at the query, it seems to me that the order by is generated from the combination of order by from inner queries (participating in cross join). The first select statement is used under inner join (for collection navigation include), therefore it should be `ORDER BY [t].[CustomerID], [t0].[CustomerID]` as in first query. (before go statement)

smitpatel commented on src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/IncludeExpressionVisitor.cs:
```diff
@@ -378,7 +378,10 @@ var newJoinExpression
         {
-            foreach (var ordering in innerJoinSelectExpression.OrderBy)
+            var orderings = innerJoinSelectExpression.OrderBy.ToList();
+            innerJoinSelectExpression.ClearOrderBy();
```
We cannot clear Order by here since they are required for the query which contains Top or Offset. See line 418.

Tasteful replied:
I have now added so the clear/adding orderby is in condition

smitpatel commented:
Potential fix for #5038 & #4695
